5637580bbf71eca6468cd4a5e15c72860928ec92
src/main/java/com/yelp/nrtsearch/server/grpc/DeadlineUtils.java
--- 
+++ 
@@ -15,6 +15,7 @@
  */
 package com.yelp.nrtsearch.server.grpc;
 
+import com.yelp.nrtsearch.server.grpc.SearchResponse.Diagnostics;
 import com.yelp.nrtsearch.server.monitoring.DeadlineMetrics;
 import io.grpc.Context;
 import io.grpc.Deadline;
@@ -54,4 +55,21 @@
       }
     }
   }
+
+  public static void checkDeadline(
+      String message, Diagnostics.Builder diagnostics, String operation) {
+    if (cancellationEnabled) {
+      Deadline deadline = Context.current().getDeadline();
+      if (deadline != null && deadline.isExpired()) {
+        DeadlineMetrics.nrtDeadlineCancelCount.labels(operation).inc();
+        throw Status.CANCELLED
+            .withDescription(
+                "Request deadline exceeded: "
+                    + message
+                    + ", Search Diagnostics: "
+                    + diagnostics.toString())
+            .asRuntimeException();
+      }
+    }
+  }
 }